session.go

```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
 	sess.breakCh = c
 }
 
-func (sess *session) handleRequests(reqs <-chan *gossh.Request) {
-	for req := range reqs {
-		switch req.Type {
-		case "shell", "exec":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+func (sess *session) handleRequests(ctx Context, reqs <-chan *gossh.Request) {
+	keepAliveEnabled := sess.keepAliveInterval > 0
+	lastReceived := time.Now()
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.rawCmd = payload.Value
+	var keepAliveCallback func()
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
-			}
+	var keepAliveTicker *time.Ticker
+	if keepAliveEnabled {
+		keepAliveTicker = time.NewTicker(sess.keepAliveInterval)
+		defer keepAliveTicker.Stop()
 
-			sess.handled = true
-			req.Reply(true, nil)
+		keepAliveCallback = func() {
+			lastReceived = time.Now()
+			keepAliveTicker.Reset(sess.keepAliveInterval)
+		}
 
-			go func() {
-				sess.handler(sess)
-				sess.Exit(0)
-			}()
-		case "subsystem":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+		ctx.SetValue(ContextKeyKeepAliveCallback, keepAliveCallback)
+	} else {
+		// Configure a stopped ticker to prevent `<-keepAliveTicker.C` from panicking.
+		keepAliveTicker = time.NewTicker(math.MaxInt64)
+		keepAliveTicker.Stop()
+	}
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.subsystem = payload.Value
+	for {
+		select {
+		case <-keepAliveTicker.C:
+			if lastReceived.Add(time.Duration(sess.keepAliveCountMax) * sess.keepAliveInterval).Before(time.Now()) {
+				log.Println("Keep-alive reply not received. Close down the session.")
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
+				err := sess.Close()
+				if err != nil {
+					log.Printf("Closing session failed: %v", err)
+				}
+				return
 			}
 
-			handler := sess.subsystemHandlers[payload.Value]
-			if handler == nil {
-				handler = sess.subsystemHandlers["default"]
+			log.Println("Send keep-alive request to the client")
+			// reply can be either false or true, but it always means that the client is alive
+			_, err := sess.SendRequest(keepAliveRequestType, true, nil)
```

Comment:
Since this is run synchronously, won't it hold up other SSH events? For instance, say a resize happens at the same time, that resize would be delayed by the whole RTT.

Perhaps we need a different kind of keepalive implementation? We probably can't simply do `go ...` here either since we want to ensure there's only one outbound at a time.

Reply:
Personally I don't think any added lag to the terminal is acceptable and will result in a degraded user experience. Proxied pings to my workspace take 300ms on dev.coder.com, if a keepalive+resize happen simultaneously I would expect that's around 450ms or more before the change is reflected in the terminal (an addition of 150ms is quite severe IMO).

Reply:
Actually, my math is off. It might be double: Resize from client 150ms -> KA from server 150ms -> KA Resp from client 150ms -> Resize on server + response 150ms -> Total 600ms.

Reply:
According to the source `pty-req` is also blocking, right?

```
	ok, err := s.ch.SendRequest("pty-req", true, Marshal(&req))
	if err == nil && !ok {
		err = errors.New("ssh: pty-req failed")
	}
```

so it seems that you can't intercept `pty-req` with `keepalive@openssh.com`, unless I understand it wrong. But I 
agree that aligning 2 consecutive blocking requests one after another can be concerning.

Reply:
Isn’t `pry-req` only done once at the start if the session? I think `window-change` is a more apt comparison.

Reply:
I missed that, thanks for clarification. Indeed, `window-change` is non-blocking, so it may be affected by keep-alive packets: 

```
_, err := s.ch.SendRequest("window-change", false, Marshal(&req))
return err
```